File: ClassBuilder/Builder.fs
﻿namespace ClassBuilder.Classes

open Autofac
open Caelan.Frameworks.Common.Helpers
open ClassBuilder
open ClassBuilder.Interfaces
open System
open System.Collections.ObjectModel
open System.Linq
open System.Reflection

module Builder = 
    let private assemblies = ObservableCollection<Assembly>()
    let private isMapper (t : Type) = typeof<IMapper>.IsAssignableFrom(t) && not t.IsAbstract && not t.IsInterface && not t.IsGenericTypeDefinition
    
    let private containsMapper (t : Assembly) = 
        try 
            t.GetTypes() |> Array.exists isMapper
        with _ -> false
    
    let mutable private container = 
        let cb = ContainerBuilder()
        cb.RegisterGeneric(typedefof<DefaultMapper<_, _>>).As(typedefof<IMapper<_, _>>) |> ignore
        cb.RegisterAssemblyTypes(assemblies.AsEnumerable() |> Array.ofSeq).Where(fun t -> t |> isMapper).AsImplementedInterfaces() |> ignore
        cb.Build()
    
    let RegisterMapper<'TMapper when 'TMapper :> IMapper>() = 
        let cb = ContainerBuilder()
        cb.RegisterType<'TMapper>().AsImplementedInterfaces() |> ignore
        cb.Update(container) |> ignore
    
    let internal registerAssemblies (allAssemblies : Assembly []) = 
        let cb = ContainerBuilder()
        cb.RegisterAssemblyTypes(allAssemblies).Where(fun t -> t |> isMapper).AsImplementedInterfaces() |> ignore
        cb.Update(container) |> ignore
        allAssemblies
        |> Array.collect (fun t -> t.GetReferencedAssemblies())
        |> Array.map (fun t -> try
                                    t |> Assembly.Load
                                with _ -> null)
        |> Array.filter (isNull >> not)
        |> Array.filter (assemblies.Contains >> not)
        |> Array.filter containsMapper
        |> Array.iter assemblies.Add
    
    assemblies.CollectionChanged.Add(fun t -> 
        registerAssemblies (t.NewItems.Cast<Assembly>()
                            |> Seq.filter (isNull >> not)
                            |> Array.ofSeq))
    AppDomain.CurrentDomain.GetAssemblies() |> registerAssemblies
    
    let internal getMapper<'TSource, 'TDestination>() = 
        let mutable mapper = Unchecked.defaultof<IMapper<'TSource, 'TDestination>>
        use scope = container.BeginLifetimeScope()
        container.TryResolve<IMapper<'TSource, 'TDestination>>(&mapper) |> ignore
        mapper
    
    [<Sealed>]
    type Builder<'T> internal (source) = 
        
        /// <summary>
        /// 
        /// </summary>
        member this.To<'TDestination>() = 
            let destination = Activator.CreateInstance<'TDestination>()
            
            let mapper = 
                if typeof<'TDestination>.Assembly
                   |> (isNull >> not)
                   && typeof<'TDestination>.Assembly |> (assemblies.Contains >> not)
                   && typeof<'TDestination>.Assembly |> containsMapper then assemblies.Add(typeof<'TDestination>.Assembly)
                getMapper<'T, 'TDestination>()
            (destination, mapper) |> this.To
        
        /// <summary>
        /// 
        /// </summary>
        /// <param name="mapper"></param>
        member this.To<'TDestination>(mapper : IMapper<'T, 'TDestination>) = (Activator.CreateInstance<'TDestination>(), mapper) |> this.To
        
        /// <summary>
        /// 
        /// </summary>
        /// <param name="destination"></param>
        member this.To<'TDestination>(destination : 'TDestination) = 
            let mapper = 
                if typeof<'TDestination>.Assembly
                   |> (isNull >> not)
                   && typeof<'TDestination>.Assembly |> (assemblies.Contains >> not)
                   && typeof<'TDestination>.Assembly |> containsMapper then assemblies.Add(typeof<'TDestination>.Assembly)
                getMapper<'T, 'TDestination>()
            (destination, mapper) |> this.To
        
        /// <summary>
        /// 
        /// </summary>
        /// <param name="destination"></param>
        /// <param name="mapper"></param>
        member __.To<'TDestination>(destination, mapper : IMapper<'T, 'TDestination>) = 
            match source :> obj |> Option.ofObj with
            | None -> Unchecked.defaultof<'TDestination>
            | Some(s) -> (source, destination) |> mapper.Map
    
    [<Sealed>]
    type ListBuilder<'T> internal (sourceList) = 
        
        /// <summary>
        /// 
        /// </summary>
        member this.ToList<'TDestination>() = 
            if typeof<'TDestination>.Assembly
               |> (isNull >> not)
               && typeof<'TDestination>.Assembly |> (assemblies.Contains >> not)
               && typeof<'TDestination>.Assembly |> containsMapper then assemblies.Add(typeof<'TDestination>.Assembly)
            getMapper<'T, 'TDestination>() |> this.ToList
        
        /// <summary>
        /// 
        /// </summary>
        /// <param name="mapper"></param>
        member __.ToList<'TDestination>(mapper : IMapper<'T, 'TDestination>) = sourceList |> Seq.map mapper.Map
    
    /// <summary>
    /// 
    /// </summary>
    /// <param name="source"></param>
    let Build<'T>(source : 'T) = Builder<'T>(source)
    
    /// <summary>
    /// 
    /// </summary>
    /// <param name="sourceList"></param>
    let BuildList<'T>(sourceList : seq<'T>) = ListBuilder<'T>(sourceList)

File: ClassBuilder/Classes/Builder.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Reflection;
using Autofac;
using ClassBuilder.Interfaces;

namespace ClassBuilder.Classes
{
    public static class Builder
    {
        private static readonly ObservableCollection<Assembly> Assemblies;
        private static readonly IContainer Container;

        static Builder()
        {
            Assemblies = new ObservableCollection<Assembly>();

            var cb = new ContainerBuilder();

            cb.RegisterGeneric(typeof(DefaultMapper<,>)).As(typeof(IMapper<,>));
            cb.RegisterAssemblyTypes(Assemblies.ToArray()).Where(IsMapper).AsImplementedInterfaces();

            Container = cb.Build();

            Assemblies.CollectionChanged += (sender, args) =>
            {
                RegisterAssemblies(args.NewItems.Cast<Assembly>().Where(t => t != null).ToArray());
            };

            RegisterAssemblies(AppDomain.CurrentDomain.GetAssemblies());
        }

        internal static void RegisterAssemblies(params Assembly[] assemblies)
        {
            var cb = new ContainerBuilder();
            cb.RegisterAssemblyTypes(assemblies).Where(IsMapper).AsImplementedInterfaces();
            cb.Update(Container);

            foreach (var a in assemblies.SelectMany(t => t.GetReferencedAssemblies()).Select(AssemblyLoadSecure).Where(t => t != null && !assemblies.Contains(t) && ContainsMapper(t)))
            {
                Assemblies.Add(a);
            }
        }

        internal static IMapper<TSource, TDestination> GetMapper<TSource, TDestination>()
        {
            IMapper<TSource, TDestination> mapper;

            using (var scope = Container.BeginLifetimeScope())
            {
                scope.TryResolve(out mapper);
            }

            return mapper;
        }

        private static bool IsMapper(Type t) => typeof(IMapper).IsAssignableFrom(t) && !t.IsAbstract && !t.IsInterface && !t.IsGenericTypeDefinition;

        private static Assembly AssemblyLoadSecure(AssemblyName assemblyName)
        {
            try
            {
                return Assembly.Load(assemblyName);
            }
            catch (Exception)
            {
                return null;
            }
        }
        private static bool ContainsMapper(Assembly a)
        {
            try
            {
                return a.GetTypes().Any(IsMapper);
            }
            catch (Exception)
            {
                return false;
            }
        }

        public static void RegisterMapper<TMapper>() where TMapper : IMapper
        {
            var cb = new ContainerBuilder();
            cb.RegisterType<TMapper>().AsImplementedInterfaces();
            cb.Update(Container);
        }

        public static Builder<TSource> Build<TSource>(TSource source) => new Builder<TSource>(source);

        public static ListBuilder<TSource> BuildList<TSource>(IEnumerable<TSource> sourceList) => new ListBuilder<TSource>(sourceList);
    }

    public sealed class Builder<TSource>
    {
        private readonly TSource _source;

        internal Builder(TSource source)
        {
            _source = source;
        }

        public TDestination To<TDestination>()
        {
            Builder.RegisterAssemblies(typeof(TDestination).Assembly);

            return To(Activator.CreateInstance<TDestination>(), Builder.GetMapper<TSource, TDestination>());
        }

        public TDestination To<TDestination>(TDestination destination)
        {
            Builder.RegisterAssemblies(typeof(TDestination).Assembly);

            return To(destination, Builder.GetMapper<TSource, TDestination>());
        }

        public TDestination To<TDestination>(IMapper<TSource, TDestination> mapper)
        {
            Builder.RegisterAssemblies(typeof(TDestination).Assembly);

            return To(Activator.CreateInstance<TDestination>(), mapper);
        }

        public TDestination To<TDestination>(TDestination destination, IMapper<TSource, TDestination> mapper)
        {
            return _source?.Equals(default(TSource)) ?? true ? default(TDestination) : mapper.Map(_source, destination);
        }
    }

    public sealed class ListBuilder<TSource>
    {
        private readonly IEnumerable<TSource> _sourceList;

        internal ListBuilder(IEnumerable<TSource> sourceList)
        {
            _sourceList = sourceList;
        }

        public IEnumerable<TDestination> ToList<TDestination>()
        {
            Builder.RegisterAssemblies(typeof(TDestination).Assembly);

            return ToList(Builder.GetMapper<TSource, TDestination>());
        }

        public IEnumerable<TDestination> ToList<TDestination>(IMapper<TSource, TDestination> mapper)
        {
            Builder.RegisterAssemblies(typeof(TDestination).Assembly);

            return _sourceList.Select(mapper.Map);
        }
    }
}

